plottr/plot/mpl/autoplot.py

```diff
@@ -0,0 +1,505 @@
+            'dataStructureChanged': dataStructure != self.dataStructure,
+            'dataShapesChanged': dataShapes != self.dataShapes,
+            'dataLimitsChanged': dataLimits != self.dataLimits,
+        }
+
+        self.dataType = dataType
+        self.dataStructure = dataStructure
+        self.dataShapes = dataShapes
+        self.dataLimits = dataLimits
+
+        return result
+
+    def dataIsComplex(self, dependentName: Optional[str] = None) -> bool:
+        """Determine whether our data is complex.
+        If dependent_name is not given, check all dependents, return True if any
+        of them is complex.
+        """
+        if self.data is None:
+            return False
+
+        if dependentName is None:
+            for d in self.data.dependents():
+                if np.issubsctype(self.data.data_vals(d), np.complexfloating):
+                    return True
+        else:
+            if np.issubsctype(self.data.data_vals(dependentName), np.complexfloating):
+                return True
+
+        return False
+
+    def setData(self, data: Optional[DataDictBase]) -> None:
+        """Analyses data and determines whether/what to plot.
+
+        :param data: input data
+        """
+        super().setData(data)
+
+        changes = self._analyzeData(data)
+        self.plotDataType = determinePlotDataType(data)
+
+        self._processPlotTypeOptions()
+        self._plotData()
+
+    def _processPlotTypeOptions(self) -> None:
+        """Given the current data type, figure out what the plot options are."""
+        if self.plotDataType == PlotDataType.grid2d:
+            self.plotOptionsToolBar.setAllowedPlotTypes(
+                PlotType.image, PlotType.colormesh, PlotType.scatter2d
+            )
+
+        elif self.plotDataType == PlotDataType.scatter2d:
+            self.plotOptionsToolBar.setAllowedPlotTypes(
+                PlotType.scatter2d,
+            )
+
+        elif self.plotDataType in [PlotDataType.scatter1d,
+                                   PlotDataType.line1d]:
+            self.plotOptionsToolBar.setAllowedPlotTypes(
+                PlotType.multitraces, PlotType.singletraces,
+            )
+
+        else:
+            self.plotOptionsToolBar.setAllowedPlotTypes()
+
+    @Slot(PlotType)
+    def _plotTypeFromToolBar(self, plotType: PlotType) -> None:
+        if plotType is not self.plotType:
+            self.plotType = plotType
+            self._plotData()
+
+    @Slot(bool)
+    def _complexPreferenceFromToolBar(self, magPhasePreferred: bool) -> None:
+        if magPhasePreferred:
+            self.complexPreference = ComplexRepresentation.magAndPhase
+        else:
+            self.complexPreference = ComplexRepresentation.realAndImag
+
+        self._plotData()
+
+    def _plotData(self) -> None:
+        """Plot the data using previously determined data and plot types."""
+
+        if self.plotDataType is PlotDataType.unknown:
+            logger.debug("No plotable data.")
```

Comment:
missing early return?
```suggestion
            logger.debug("No plotable data.")
            return
```